dev/AppLifecycle/Association.cpp

```diff
@@ -54,13 +54,19 @@ namespace winrt::Microsoft::ProjectReunion::implementation
         return path;
     }
 
-    std::wstring ComputeAppId()
+    std::wstring ComputeAppId(const std::wstring& customSeed)
     {
         // PRF = Project Reunion Framework.  A simple prefix to help identify ProgIds we compute here.
-        // AppId = Prefix + Hash(modulePath)
```

Comment:
>Hash(modulePath) [](start = 28, length = 16)

If I move my exe it's a totally different value?
If I unzip some app to C:\Foo and register things, then move it to C:\Bar I need to register to pick up the changed path. Is this documented?

Also, I now have old data in the registry pointing to C:\Foo with no way for it to be removed?

Is there a way to avoid such WinRot? Can you hash just the filename of the exe? Some input provided by the app?
Is a Cleanup() API to find old orphaned data and delete them possible? Or is that ambiguous with valid Win32 practices?

Reply:
This is a known limitation that we're living with.  The scenarios where this go wrong are:

1. App moves legitimately as part of some app supported scenario.  The app is responsible for correcting it's associations anyway (remove re-add later).
2. App is moved by user even though it doesn't support it.  Nothing we can do here, as the EXE path would be wrong anyway and nothing would run.
3. The fact that the App can specify the seed means it can recreate the required information for later removal using the API.

There is no solution for this type of WinRot that can be solved in this component as it's an open playground in the registry.  Ultimately the plan is to move to a more stable seed, ideally something like the publisher from a signature or the like.  For now this is what we've decided to do.

Reply:
>There is no solution for this type of WinRot that can be solved in this component

One option is a "LifetimeArtifact" style solution, similar to what @jonwis had suggested for a similar problem for Dynamic Dependencies. To handle when a definition's created and the software goes away without deleting the definition (dev forgets, app's deleted rather than 'properly' uninstalled, if an app moves it needs to register but not worry about deregistering previous definition, etc).

For instance, to detect WinRot when using a protocol

1. RegisterForProtocolActivation(...LifetimeArtifact="c:\foo\bar.exe")
2. When writing the activation info also RegWrite("LifetimeArtifact", REG_SZ, "c:\foo\bar.exe")
3. When the protocol is activated and the definition is retrieved from the registry, also retrieve LifetimeArtifact and if (!fileexists(filename)) then consider the definition 'expired' -- Unregister it (best-effort) and act as if the protocol registration didn't exist

Another option:
1. RegisterForProtocolActivation(...LifetimeArtifact="c:\foo\bar.exe")
2. When writing the activation info also write info to a central location (HKCU\SOFTWARE\Something, the ProjectReunionMain package's ApplicationData, etc)
3. Have a recurring job (e.g. BackgroundTask) enumerate this list, verify the LifetimeArtifact exists and if not call Unregister

DynamicDependencies supports this sort of garbage collection. MddTryCreatePackageDependency takes a filename or regkey. On use (MddAddDynamicDependency) it checks if a definition's 'expired', and an upcoming addition will regularly check for and remove obsolete DynamicDependencyLifetimeManager packages.

There may be other options. Bottom lie is to tie the definition to some durable artifact and when that goes away, the definition goes away. For Protocol/etc activation the executable can often be a good LifetimeArtifact as it's typically the target for a filetype or protocol activation. After all if the target exe goes away the definition likely isn't needed anymore either.

Question is how important is it to prevent WinRot.

IMO it's not significant for this PR _other than_ its affect on the API.